### PR TITLE
Expand shell wildcards in --paths for all platforms

### LINKED ISSUE #438: "paths" & "base-paths" ignored in defaults.yaml
```yaml
{
  "base-paths": "src/*/py"
  "paths": "src/*/py"
  "": {
    "base-paths": "src/*/py"
    "paths": "src/*/py"
  }
  "build": {
    "paths": "src/*/py"
    "symlink-install": true
    "cmake-args": "-DCMAKE_BUILD_TYPE=Debug"
  }
  "list": {
    "base-paths": "src/*/py"
    "paths": "src/*/py"
  }
}
```

None of the "paths" or "base-paths" verbs are respected by Colcon.  
`colcon list --paths-only --paths src/* src/*/py` shows expected output  
`colcon list --paths-only` does not

### PR DESCRIPTION
**Pull request by christophebedard**
Fixes #438

If the `--paths` value is not provided through the shell, then shell wildcards won't be expanded. This is the case when using a `defaults.yaml` file.

To cover this case, expand wildcards on all platforms, not just Windows. I renamed and made `expand_wildcards` part of the public API so that it can be used elsewhere as well.

I can't think of any downsides to doing this, and being able to use wildcards in a `defaults.yaml` file does seem like a very normal use case.

Signed-off-by: Christophe Bedard <bedard.christophe@gmail.com>

### CODE REVIEW

**Comment by christophebedard** [comment]
I think it's also important to point out that it doesn't support double asterisks (`**`). Without `recursive=True` for `glob`, it is pretty much treated as a simple `*`: https://docs.python.org/3/library/glob.html#glob.glob

Maybe we should support that? The call to `glob` does take more time with `recursive=True` of course, but if people using `**` they should be expecting that anyway.

**Comment by christophebedard** [comment]
I renamed and made `expand_wildcards` part of the public API so that it can be used elsewhere, e.g. by `colcon-recursive-crawl` for `--base-paths` https://github.com/colcon/colcon-recursive-crawl/pull/22

**Comment by christophebedard** [comment]
@cottsay @sloretz @nuclearsandwich friendly ping. Could this get a review?

**Review by cottsay — COMMENTED** [review]
I think that consistent behavior across shell and non-shell invocation, as well as between platforms is desirable. So :+1: for moving forward here.

That said, I'm lukewarm about the existing implementation, and I'm not sure it would be a good idea to expand use of it as-is. In particular, specifically ignoring matching files is convenient, but deviates from what a typical shell would do. Also, as far as I can tell, a given pattern which has no matches is silently skipped by this implementation, where a typical shell would simply forward it as-is.

Another thought when promoting this functionality in the public API is whether tapping into argparse to resolve the paths might be a cleaner way to use the functionality easily across other path arguments in colcon.

I mocked up something that might do the trick: https://gist.github.com/cottsay/6a153ed69a3471e739cf911151fb63de

Just drop `action=WildcardPathAction` in the call to `add_argument` to use one of the classes. I included a class which implements the aforementioned file match filtering, but again, I'm still not sure it's actually a good idea.

**Comment by christophebedard** [comment]
> a given pattern which has no matches is silently skipped by this implementation, where a typical shell would simply forward it as-is.

If a pattern yields no matches, then there isn't really anything else that can be done with it, is there? I don't think the subsequent code/logic can do anything more with it.

**Comment by cottsay** [comment]
> If a pattern yields no matches, then there isn't really anything else that can be done with it, is there? I don't think the subsequent code/logic can do anything more with it.

My comment was maybe a little premature - I hadn't actually tried this out and only inferred the behavior from the code. The existing implementation also seems to ignore non-existing paths without wildcards, which I hadn't expected. I'd guess this changes my interpretation of what purpose `--paths` aims for. It appears to be a list of possible search patterns rather than a literal list of paths to be searched.

Sorry about that - I should have done the due diligence of understanding the full behavior of the argument before critiquing it so.

Since this wasn't very clear in the code, and I don't expect it to be nearly as widely used across colcon as I previously believed, I think moving forward as-is would be fine. One request though - could you rename the (now public) function and/or add a comment explaining the nuances of this behavior? I would expect a function called `expand_wildcards` to match files and directories equally.

**Comment by christophebedard** [comment]
@cottsay that's fair. I renamed it to `expand_dir_wildcards` and documented its behaviour in a1ebd5fe686012755f7e03d9b32879d6cc76e2b4

**Comment by christophebedard** [comment]
I also rebased on `master` to get the CI changes.

**Review by cottsay — APPROVED** [review]

**Comment by christophebedard** [comment]
@cottsay thanks for the review! Does it need a second review, or can it be merged? Then https://github.com/colcon/colcon-recursive-crawl/pull/22 should be a trivial change after this PR, but `colcon-core` will need to be bumped and released (e.g., 0.6.2) for `colcon-recursive-call` to use this new public function.

**Pull request merged**
